File: app/src/main/java/io/github/freewebmovement/igniter/ui/exempt/fragment/ExemptAppFragment.java
package io.github.freewebmovement.igniter.ui.exempt.fragment;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.menu.MenuBuilder;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.SwitchCompat;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.snackbar.Snackbar;

import java.util.List;

import io.github.freewebmovement.igniter.IgniterApplication;
import io.github.freewebmovement.igniter.R;
import io.github.freewebmovement.igniter.common.app.BaseFragment;
import io.github.freewebmovement.igniter.common.dialog.LoadingDialog;
import io.github.freewebmovement.igniter.ui.exempt.adapter.AppInfoAdapter;
import io.github.freewebmovement.igniter.ui.exempt.contract.ExemptAppContract;
import io.github.freewebmovement.igniter.ui.exempt.data.AppInfo;

public class ExemptAppFragment extends BaseFragment implements ExemptAppContract.View {
    public static final String TAG = "ExemptAppFragment";
    private ExemptAppContract.Presenter mPresenter;
    private Toolbar mTopBar;
    private RecyclerView mAppRv;
    private AppInfoAdapter mAppInfoAdapter;
    private LoadingDialog mLoadingDialog;
    IgniterApplication app;
    public ExemptAppFragment() {
        // Required empty public constructor
    }

    public static ExemptAppFragment newInstance() {
        return new ExemptAppFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_exempt_app, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        app = IgniterApplication.getApplication();
        super.onViewCreated(view, savedInstanceState);
        findViews();
        initViews();
        initListeners();
        mPresenter.start();
    }

    private void findViews() {
        mTopBar = findViewById(R.id.exemptAppTopBar);
        mAppRv = findViewById(R.id.exemptAppRv);
    }

    private void initViews() {
        FragmentActivity activity = getActivity();
        if (activity instanceof AppCompatActivity) {
            ((AppCompatActivity) activity).setSupportActionBar(mTopBar);
            setHasOptionsMenu(true);
        }
        mAppInfoAdapter = new AppInfoAdapter();
        mAppRv.setAdapter(mAppInfoAdapter);
        mAppRv.addItemDecoration(new DividerItemDecoration(mContext, LinearLayoutManager.VERTICAL));
    }

    private void initListeners() {
        mAppInfoAdapter.setOnItemOperationListener((exempt, appInfo, position) -> mPresenter.updateAppInfo(appInfo, position, exempt));
    }

    @SuppressLint("RestrictedApi")
    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        menu.clear();
        inflater.inflate(R.menu.menu_exempt_app, menu);
        if(menu instanceof MenuBuilder){
            MenuBuilder m = (MenuBuilder) menu;
            m.setOptionalIconsVisible(true);
        }

        initHideSystemSwitch(menu);

        MenuItem item = menu.findItem(R.id.action_search_app);
        SearchView searchView = null;
        if (item != null) {
            searchView = (SearchView) item.getActionView();
        }
        if (searchView != null) {
            searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
                @Override
                public boolean onQueryTextSubmit(String s) {
                    return false;
                }

                @Override
                public boolean onQueryTextChange(String s) {
                    mPresenter.filterAppsByName(s);
                    return true;
                }
            });
        }
    }

    public void initHideSystemSwitch(Menu menu) {
        MenuItem item = (MenuItem) menu.findItem(R.id.hide_system_apps);
        if (item != null) {
            SwitchCompat switchCompat = (SwitchCompat)item.getActionView();
            switchCompat.setChecked(app.trojanPreferences.getShowSystemApps());
            switchCompat.setOnCheckedChangeListener((buttonView, isChecked) -> {
                app.trojanPreferences.setShowSystemApps(isChecked);
                mPresenter.start();
            });
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_save_exempt_apps) {
            mPresenter.saveExemptAppInfoList();
            return true;
        }
        return false;
    }

    @Override
    public void showSaveSuccess() {
        Snackbar.make(mRootView, R.string.common_save_success, Snackbar.LENGTH_SHORT).setAction(R.string.exempt_app_exit,
                v -> mPresenter.exit()).show();
    }

    @Override
    public void showExitConfirm() {
        new AlertDialog.Builder(mContext)
                .setTitle(R.string.common_alert)
                .setMessage(R.string.exempt_app_exit_without_saving_confirm)
                .setNegativeButton(R.string.common_cancel, (dialog, which) -> dialog.dismiss())
                .setPositiveButton(R.string.common_confirm, (dialog, which) -> {
                    dialog.dismiss();
                    mPresenter.exit();
                }).create().show();
    }

    @Override
    public void showAppList(final List<AppInfo> appInfoList) {
        mAppInfoAdapter.refreshData(appInfoList);
    }

    @Override
    public void showLoading() {
        if (mLoadingDialog == null) {
            mLoadingDialog = new LoadingDialog(requireContext());
            mLoadingDialog.setMsg(getString(R.string.exempt_app_loading_tip));
        }
        mLoadingDialog.show();
    }

    @Override
    public void dismissLoading() {
        if (mLoadingDialog != null && mLoadingDialog.isShowing()) {
            mLoadingDialog.dismiss();
        }
    }

    @Override
    public void exit(boolean configurationChanged) {
        Activity activity = getActivity();
        if (activity != null) {
            activity.setResult(configurationChanged ? Activity.RESULT_OK : Activity.RESULT_CANCELED);
            activity.finish();
        }
    }

    @Override
    public void setPresenter(ExemptAppContract.Presenter presenter) {
        mPresenter = presenter;
    }
}

File: app/src/main/java/io/github/freewebmovement/igniter/activities/exempt/fragment/ExemptAppFragment.kt
package io.github.freewebmovement.igniter.activities.exempt.fragment

import android.annotation.SuppressLint
import android.app.Activity
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.snackbar.Snackbar
import io.github.freewebmovement.igniter.IgniterApplication
import io.github.freewebmovement.igniter.R
import io.github.freewebmovement.igniter.activities.MainActivity
import io.github.freewebmovement.igniter.activities.exempt.adapter.AppInfoAdapter
import io.github.freewebmovement.igniter.activities.exempt.contract.ExemptAppContract
import io.github.freewebmovement.igniter.common.app.BaseFragment
import io.github.freewebmovement.igniter.common.dialog.AppSheet
import io.github.freewebmovement.igniter.databinding.FragmentExemptAppBinding
import io.github.freewebmovement.igniter.persistence.data.AppInfo

class ExemptAppFragment : BaseFragment(), ExemptAppContract.View {
    private var mPresenter: ExemptAppContract.Presenter? = null
    private var _binding: FragmentExemptAppBinding? = null
    private val binding get() = _binding!!
    private lateinit var mAppInfoAdapter: AppInfoAdapter
    private val app: IgniterApplication
        get() = IgniterApplication.getApplication()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        _binding = FragmentExemptAppBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
        initListeners()
        initTabs()
        mPresenter!!.start()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun initViews() {
        mAppInfoAdapter = AppInfoAdapter()
        binding.exemptAppRv.adapter = mAppInfoAdapter
        binding.exemptAppRv.addItemDecoration(DividerItemDecoration(requireContext(), LinearLayoutManager.VERTICAL))
    }

    private fun initTabs() {
        binding.proxyTabNormal.isSelected = true
        binding.proxyTabSystem.isSelected = false
        binding.proxyTabNormal.setOnClickListener {
            binding.proxyTabNormal.isSelected = true
            binding.proxyTabSystem.isSelected = false
            mPresenter!!.switchTab(false)
        }
        binding.proxyTabSystem.setOnClickListener {
            binding.proxyTabNormal.isSelected = false
            binding.proxyTabSystem.isSelected = true
            mPresenter!!.switchTab(true)
        }
    }

    private fun initListeners() {
        mAppInfoAdapter.setOnItemOperationListener(object : AppInfoAdapter.OnItemOperationListener {
            override fun onToggle(enabled: Boolean, appInfo: AppInfo, position: Int) {
                mPresenter!!.updateAppInfo(appInfo, position, enabled)
            }
        })

        binding.searchExemptAppEt.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                mPresenter!!.filterAppsByName(s?.toString().orEmpty())
            }
        })

        binding.btnSelectAll.setOnClickListener { mPresenter!!.selectAll() }
        binding.btnDeselectAll.setOnClickListener { mPresenter!!.deselectAll() }
        binding.btnSaveProxyApps.setOnClickListener { mPresenter!!.saveExemptAppInfoList() }
    }

    override fun showSaveSuccess() {
        Snackbar.make(binding.root, R.string.common_save_success, Snackbar.LENGTH_SHORT).show()
    }

    override fun showExitConfirm() {
        AppSheet.builder(this)
            .setTitle(R.string.common_alert)
            .setMessage(R.string.exempt_app_exit_without_saving_confirm)
            .setNegativeButton(R.string.common_cancel, null)
            .setPositiveButton(R.string.common_confirm) {
                mPresenter!!.exit()
            }
            .show()
    }

    override fun showAppList(appInfoList: List<AppInfo>) {
        mAppInfoAdapter.refreshData(appInfoList)
    }

    override fun showLoading() {
        AppSheet.builder(this).showLoading(getString(R.string.exempt_app_loading_tip))
    }

    override fun dismissLoading() {
        AppSheet.dismissActive()
    }

    override fun exit(configurationChanged: Boolean) {
        val activity = activity
        if (activity is MainActivity) {
            activity.openHomeTab()
            return
        }
        if (activity != null) {
            activity.setResult(if (configurationChanged) Activity.RESULT_OK else Activity.RESULT_CANCELED)
            activity.finish()
        }
    }

    override fun setPresenter(presenter: ExemptAppContract.Presenter) {
        mPresenter = presenter
    }

    companion object {
        const val TAG = "ExemptAppFragment"

        @JvmStatic
        fun newInstance(): ExemptAppFragment {
            return ExemptAppFragment()
        }
    }
}
